zerver/actions/presence.py

```diff
@@ -162,8 +175,65 @@ def do_update_user_presence(
-    if len(update_fields) > 0:
-        presence.save(update_fields=update_fields)
+
+    # WARNING: Delicate, performance-sensitive block.
+
+    # It's time to determine last_update_id and update the presence object in the database.
+    # This briefly takes the crucial lock on the PresenceSequence row for the user's realm.
+    # We're doing this in a single SQL query to avoid any Python-level overhead. We're also
+    # intentionally doing this at the very end of the function, at the last step before the
+    # transaction commits. This ensures the lock is held for the shortest time possible.
+
+    # Equivalent Python code:
+    # if created or len(update_fields) > 0:
+    #     presence_sequence = PresenceSequence.objects.select_for_update().get(realm_id=user_profile.realm_id)
+    #     new_last_update_id = presence_sequence.last_update_id + 1
+    #     presence_sequence.last_update_id = new_last_update_id
+    #     if created:
+    #         presence.last_update_id = new_last_update_id
+    #         presence.save()
+    #     elif len(update_fields) > 0:
+    #         presence.last_update_id = new_last_update_id
+    #         presence.save(update_fields=[*update_fields, "last_update_id"])
+    #     presence_sequence.save(update_fields=["last_update_id"])
+    # But let's do it in a single, direct SQL query instead.
+
+    if created or len(update_fields) > 0:
+        query = sql.SQL("""
+            WITH new_last_update_id AS (
+                UPDATE zerver_presencesequence
+                SET last_update_id = last_update_id + 1
+                WHERE realm_id = {realm_id}
+                RETURNING last_update_id
+            )
+        """).format(realm_id=sql.Literal(user_profile.realm_id))
+
+        if created:
+            query += sql.SQL("""
+                INSERT INTO zerver_userpresence (user_profile_id, last_active_time, last_connected_time, realm_id, last_update_id)
+                VALUES ({user_profile_id}, {last_active_time}, {last_connected_time}, {realm_id}, (SELECT last_update_id FROM new_last_update_id))
+            """).format(
+                user_profile_id=sql.Literal(user_profile.id),
+                last_active_time=sql.Literal(presence.last_active_time),
+                last_connected_time=sql.Literal(presence.last_connected_time),
+                realm_id=sql.Literal(user_profile.realm_id),
+            )
+        else:
+            assert len(update_fields) > 0
+            update_fields_segment = sql.SQL(", ").join(
+                sql.SQL("{field} = {value}  ").format(
+                    field=sql.Identifier(field), value=sql.Literal(getattr(presence, field))
+                )
+                for field in update_fields
+            )
+            query += sql.SQL("""
+                UPDATE zerver_userpresence
+                SET {update_fields}, last_update_id = (SELECT last_update_id FROM new_last_update_id)
+                WHERE id = {presence_id}
```

Comment:
This is the locking in `do_update_user_presence`, flagging it here for clarity. Idea is we do the `UPDATE` to the `PresenceSequence` in a CTE in the same query as the `UserPresence` update, fetching and using the incremented `last_update_id` for the `UserPresence.last_update_id`. The `UPDATE` is what acquires the hot lock.

Here are example queries generate by this:
- Creation of presence:

```sql
            WITH new_last_update_id AS (
                UPDATE zerver_presencesequence
                SET last_update_id = last_update_id + 1
                WHERE realm_id = 2
                RETURNING last_update_id
            )
        
                INSERT INTO zerver_userpresence (user_profile_id, last_active_time, last_connected_time, realm_id, last_update_id)
                VALUES (6, NULL, '2024-05-07T21:22:28+00:00'::timestamptz, 2, (SELECT last_update_id FROM new_last_update_id))

```

- Update:
    
    ```sql
                WITH new_last_update_id AS (
                    UPDATE zerver_presencesequence
                    SET last_update_id = last_update_id + 1
                    WHERE realm_id = 2
                    RETURNING last_update_id
                )
            
                    UPDATE zerver_userpresence
                    SET "last_connected_time" = '2024-05-07T21:22:30+00:00'::timestamptz, "last_active_time" = '2024-05-07T21:22:30+00:00'::timestamptz, last_update_id = (SELECT last_update_id FROM new_last_update_id)
                    WHERE id = 17
    ```

Reply:
This looks good to me but maybe `SET {update_fields}` should be `SET {update_fields_segment}` rather than renaming it in the `.format` below, just for readability.

Reply:
This is done. Leaving the thread unresolved since it's useful highlighting this important locking mechanism for anyone else looking at the PR, with some context in my original comment above.